components/rpc/invoker/mosn/channel/xchannel.go

```diff
@@ -93,8 +97,104 @@ type xChannel struct {
 	pool  *connPool
 }
 
-// Do is handle RPCRequest to RPCResponse
-func (m *xChannel) Do(req *rpc.RPCRequest) (*rpc.RPCResponse, error) {
+// InvokeWithTargetAddress send request to specific provider address
+func (m *xChannel) InvokeWithTargetAddress(req *rpc.RPCRequest) (*rpc.RPCResponse, error) {
+	// 1. context.WithTimeout
+	timeout := time.Duration(req.Timeout) * time.Millisecond
+	ctx, cancel := context.WithTimeout(req.Ctx, timeout)
+	defer cancel()
+
+	// 2. get connection with specific address
+	conn, err := net.Dial("tcp", req.Header[rpc.TargetAddress][0])
```

Comment:
This design has several shortcomings:
1. It uses short connection instead of long connection, and  it's "one connection per request" model, which is definitely a performance-killer . 10000 requests will create and then close 10000 connections. 
This "direct-url" feature should be more generic, and designed for production. For example, sofa-rpc has `direct-url` feature and users might use it in production. See https://www.sofastack.tech/projects/sofa-rpc/peer-to-peer/

2. It bypasses mosn, which means the filters don't work. For example, the filters add tracing fields to the request or the tcpdump filter don't work, which makes debug harder

So, can we implement it like the existing code?

Reply:
1.  you can refer https://github.com/mosn/layotto/issues/705 , if you want use direct-url, you should change the config file.The address may be different for each request, no need to do conn pool。
2. Why should all the capabilities be coupled to the mosn

So, This capability is designed for the user to specify the ip of the target in the request. It is to facilitate the user to debug. If the user wants the direct-url capability, then it needs to modify the configuration file.

Reply:
2 is not a big problem, while 1 is